### PR TITLE
Add a 'Batch': collects non-transactional saves / deletes.

### LINKED ISSUE #421: Batch dataset.put and dataset.delete.
Like dataset.get, there should be a dataset.delete and dataset.put that accepts a list of keys / entities.

### PR DESCRIPTION
**Pull request by tseaver**
Closes #421.

### CODE REVIEW

**Comment by tseaver** [comment]
`Batch` largely patterend after `Transaction`:  we could maybe even derive `Transation` from batch.

Note that I dropped the "update auto_id keys" bits on the floor:  we can add them if they are important for the usecase.

File: gcloud/datastore/batch.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        :raises: ValueError if entity has no key assigned.
+        """
+        if entity.key is None:
+            raise ValueError("Entity must have a key")
+
+        key_pb = entity.key.to_protobuf()
+        properties = dict(entity)
+        exclude = tuple(entity.exclude_from_indexes)
+
+        self.connection.save_entity(
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: ISTM that the `Batch` should not make an RPC until all the entities are added to it.

File: gcloud/datastore/batch.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        :type key: :class:`gcloud.datastore.key.Key`
+        :param key: the key to be deleted.
+
+        :raises: ValueError if key is not complete.
+        """
+        if key.is_partial:
+            raise ValueError("Key must be complete")
+
+        key_pb = key.to_protobuf()
+        self.connection.delete_entities(
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: Ditto on RPC.

File: gcloud/datastore/batch.py (hunk 1 above)
Comment (tseaver) [inline gcloud/datastore/batch.py]: We pass the 'mutation' object in:  in that caes, as when there is a transaction active, the connection does not do the RPC:  it merely does the "copy to mutation" stuff.

One might arguably factor out the "copy to mutation" (into `helpers'?) from `Connection.save_entity` and `Connection.delete_entities`.

Comment (dhermes) [inline gcloud/datastore/batch.py]: Indeed. I have some other ideas relating to the desired `put()` and `delete()` methods  discussed in #421.

**Comment by dhermes** [comment]
It's pleasantly surprising how simple the tweaks to `Connection` were!

I have some other ideas floating around that relate to this (including #495). Could we jump on a video chat tomorrow AM to discuss?

-----

I was initially thrown off by the `Batch` concept since I am used to a strict distinction of `ndb.put_multi` and `ndb.delete_multi`. However, [`Mutation`][1] is happy to accommodate all.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-datastore/blob/21e029b2657ebb15fef5429c90c4315326855fd1/proto/datastore_v1.proto#L384

File: gcloud/datastore/batch.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                 are not set.
+        """
+        self._connection = connection or _implicit_environ.CONNECTION
+        self._dataset_id = dataset_id or _implicit_environ.DATASET_ID
+
+        if self._connection is None or self._dataset_id is None:
+            raise ValueError('A batch must have a connection and '
+                             'a dataset ID set.')
+
+        self._id = None
```
Comment (dhermes) [inline gcloud/datastore/batch.py]: What does this do?

File: gcloud/datastore/connection.py [hunk 4]
```diff
@@ -421,7 +421,7 @@ def allocate_ids(self, dataset_id, key_pbs):
         return list(response.key)
 
     def save_entity(self, dataset_id, key_pb, properties,
```
Comment (dhermes) [inline gcloud/datastore/connection.py]: ISTM that long term this method can be scrapped. `Connection.commit` is the only thing that maps to an RPC. All this method needs to do is update a mutation.

File: gcloud/datastore/connection.py [hunk 5]
```diff
@@ -493,7 +504,7 @@ def save_entity(self, dataset_id, key_pb, properties,
 
         return False, None
 
-    def delete_entities(self, dataset_id, key_pbs):
+    def delete_entities(self, dataset_id, key_pbs, mutation=None):
```
Comment (dhermes) [inline gcloud/datastore/connection.py]: Ditto here. We can hopefully long term just remove `mutation` from `Connection` and have functions like `add_deletes_to_mutation`.

**Comment by dhermes** [comment]
LGTM save for removing `self._id` (a remnant of `Transaction`).

I'll write up a bug outlining a "migration" path based off #495

File: gcloud/datastore/batch.py (hunk 3 above)
Comment (tseaver) [inline gcloud/datastore/batch.py]: It is probably a copy-paste fossil.

File: gcloud/datastore/connection.py (hunk 4 above)
Comment (tseaver) [inline gcloud/datastore/connection.py]: `allocate_ids` is an RPC.

Comment (dhermes) [inline gcloud/datastore/connection.py]: This was in reference to `save_entity`. I am in the split view on GitHub. Does this show up elsewhere for you?

**Comment by tseaver** [comment]
I have changes stashed which add the "insert_auto_id*" fixups to `Batch`.  Should I push them, or defer to another PR?

File: gcloud/datastore/connection.py (hunk 4 above)
Comment (tseaver) [inline gcloud/datastore/connection.py]: It showed up anchored to line 420, which is in the middle of `allocate_ids`.

**Comment by dhermes** [comment]
Let's defer to another PR so we can ask Patrick about it without lots of distracting other code.

File: gcloud/datastore/connection.py (hunk 4 above)
Comment (tseaver) [inline gcloud/datastore/connection.py]: That was on the "Conversation" tab -- looks OK on the "Unified" view of the "Files Changed" tab.

**Pull request merged**